src/NServiceBus.Core/Routing/MessageEndpointMappingCollection.cs

```diff
@@ -180,13 +182,31 @@ internal void Apply(Publishers publishers, UnicastRoutingTable unicastRoutingTab
                         return;
                     }
                     var canonicalForm = makeCanonicalAddress(endpointAddress);
-                    routeTableEntries[type] = new RouteTableEntry(type, UnicastRoute.CreateFromPhysicalAddress(canonicalForm));
-                    publisherTableEntries[type] = new PublisherTableEntry(type, PublisherAddress.CreateFromPhysicalAddresses(canonicalForm));
+                    var baseTypes = allMessageTypes.Where(candidate => candidate != type && candidate.IsAssignableFrom(type)).ToArray();
+
+                    RegisterMessageRoute(type, canonicalForm, routeTableEntries, baseTypes);
+                    RegisterEventRoute(type, canonicalForm, publisherTableEntries, baseTypes);
                 });
             }
 
-            publishers.AddOrReplacePublishers("MessageEndpointMappings", publisherTableEntries.Values.ToList());
+            publishers.AddOrReplacePublishers("MessageEndpointMappings", publisherTableEntries);
             unicastRoutingTable.AddOrReplaceRoutes("MessageEndpointMappings", routeTableEntries.Values.ToList());
         }
+
+        static void RegisterEventRoute(Type mappedType, string address, List<PublisherTableEntry> publisherTableEntries, IEnumerable<Type> baseTypes)
+        {
+            var publisherAddress = PublisherAddress.CreateFromPhysicalAddresses(address);
+            publisherTableEntries.AddRange(baseTypes.Select(derivedType => new PublisherTableEntry(derivedType, publisherAddress)));
+            publisherTableEntries.Add(new PublisherTableEntry(mappedType, publisherAddress));
+        }
+
+        static void RegisterMessageRoute(Type mappedType, string address, Dictionary<Type, RouteTableEntry> routeTableEntries, IEnumerable<Type> baseTypes)
+        {
+            foreach (var baseType in baseTypes)
+            {
+                routeTableEntries[baseType] = new RouteTableEntry(baseType, UnicastRoute.CreateFromPhysicalAddress(address));
+            }
+            routeTableEntries[mappedType] = new RouteTableEntry(mappedType, UnicastRoute.CreateFromPhysicalAddress(address));
```

Comment:
Could these two `UnicastRoute.CreateFromPhysicalAddress(address)` calls be treated like the 'publishAddress` above and call it once and use the result?